### PR TITLE
Fairshare for multisched

### PR DESCRIPTION
**Pull request by bhroam**
#### Bug/feature Description
This is interface 10 of Multisched
This adds the ability to run pbsfs with schedulers other than the default.  It also adds PTL support for the same.

#### Affected Platform(s)
* All Platforms

#### Cause / Analysis / Design
* EDD: https://pbspro.atlassian.net/wiki/spaces/PD/pages/50947131/PP-337+Multiple+schedulers+servicing+the+PBS+cluster
* Forum Post: http://community.pbspro.org/t/pp-337-multiple-schedulers-servicing-the-pbs-cluster/470/94

#### Solution Description
* There is a new option to pbsfs which takes the scheduler name.  When this option is given to pbsfs, it will contact the server and get the scheduler's sched_priv directory.  It then opens the usage file for that scheduler.

I added PTL support for the new option.  If the scheduler is not the default scheduler, the new option is given for all supported pbsfs PTL functionality.

I also ran into a couple of multisched PTL bugs which I fixed.  I also did a little refactoring to make things more consistent with the existing PTL framework.

#### Checklist:

- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [X] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [X] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_multi_sched.py [hunk 1]
```diff
@@ -579,3 +579,93 @@ def test_run_limts_per_scheduler(self):
...
+                         recursive=True, sudo=True)
+        self.setup_sc3()
+        self.scheds['sc3'].add_to_resource_group(TEST_USER, 10, 'root', 20)
+        self.scheds['sc3'].holidays_set_year(new_year="3000")
+
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'sched_priv': new_sched_priv}, id='sc3')
+
+        n = self.scheds['sc3'].fairshare_tree.get_node(id=10)
+        print n
```
Comment (latha-subramanian) [inline test/tests/functional/pbs_multi_sched.py]: Please remove the debug print

File: test/tests/functional/pbs_multi_sched.py [hunk 2]
```diff
@@ -579,3 +579,93 @@ def test_run_limts_per_scheduler(self):
...
+        self.assertFalse(n)
+
+        y = self.scheds['sc3'].holidays_get_year()
+        self.assertNotEquals(y, "3000")
+
+        # clean up
+        self.du.rm(path=new_sched_priv, sudo=True, recursive=True, force=True)
```
Comment (latha-subramanian) [inline test/tests/functional/pbs_multi_sched.py]: revert_to_defaults will do the cleanup, at the end of the test. Why is this required?

File: src/scheduler/pbsfs.c [hunk 3]
```diff
@@ -98,14 +99,14 @@ main(int argc, char *argv[])
...
-	/*the real deal or output version and exit?*/
+	/* the real deal or output version and exit? */
 	execution_mode(argc, argv);
 	set_msgdaemonname("pbsfs");
 
 	if (pbs_loadconf(0) <= 0)
 		exit(1);
 
-	while ((c = getopt(argc, argv, "sgptdcae-:")) != -1)
+	while ((c = getopt(argc, argv, "sgptdcaeI:-:")) != -1)
```
[the remark this answers is not in the document]
Comment (mike0042) [reply src/scheduler/pbsfs.c]: Yes, it's required for parameters that take arguments. The argument for -I is the name of the scheduler instance.

File: src/scheduler/pbsfs.c [hunk 4]
```diff
@@ -88,7 +88,8 @@ static void print_fairshare(group_info *root, int level);
 int
 main(int argc, char *argv[])
 {
-	char path_buf[256];
+	char path_buf[256] = {0};
+	char sched_name[PBS_MAXSCHEDNAME + 1] = {0};
```
Comment (mike0042) [inline src/scheduler/pbsfs.c]: Maybe you want to initialize this to "default" so it will contain a valid scheduler name when -I isn't specified?

File: src/scheduler/pbsfs.c [hunk 5]
```diff
@@ -131,6 +132,10 @@ main(int argc, char *argv[])
 			case 'e':
 				flags = FS_TRIM_TREE | FS_WRITE_FILE;
 				break;
+			case 'I':
+				strncpy(sched_name, optarg, sizeof(sched_name));
+				sched_name[PBS_MAXSCHEDNAME] = '\0';
```
Comment (mike0042) [inline src/scheduler/pbsfs.c]: Probably not overly concerned about performance here. You could use snprintf() or just leave it as-is.

File: src/scheduler/pbsfs.c [hunk 6]
```diff
@@ -141,7 +146,7 @@ main(int argc, char *argv[])
 		exit(1);
 	}
 	if ((flags & (FS_PRINT | FS_PRINT_TREE)) && (argc - optind) != 0) {
-		fprintf(stderr, "Usage: pbsfs -[ptdgcs]\n");
+		fprintf(stderr, "Usage: pbsfs -[ptdgcsI]\n");
```
Comment (mike0042) [inline src/scheduler/pbsfs.c]: Usage: pbsfs -[ptdgcs] [-I scheduler]

File: src/scheduler/pbsfs.c [hunk 7]
```diff
@@ -188,17 +228,17 @@ main(int argc, char *argv[])
 	else if (flags & FS_DECAY)
 		decay_fairshare_tree(conf.fairshare->root);
 	else if (flags & (FS_GET | FS_SET | FS_ADD | FS_COMP)) {
-		ginfo = find_group_info(argv[2], conf.fairshare->root);
+		ginfo = find_group_info(argv[optind], conf.fairshare->root);
```
Comment (mike0042) [inline src/scheduler/pbsfs.c]: We haven't validated optind in the case this is FS_ADD. The rest are validated above.

File: test/tests/functional/pbs_multi_sched.py [hunk 8]
```diff
@@ -579,3 +579,93 @@ def test_run_limts_per_scheduler(self):
         self.server.expect(JOB, {'job_state': 'Q'}, id=jid4)
         jc = "Not Running: User has reached server running job limit."
         self.server.expect(JOB, {'comment': jc}, id=jid4)
+
+    def test_multi_fairshare(self):
```
Comment (mike0042) [inline test/tests/functional/pbs_multi_sched.py]: Do you want to add another case where we pass an invalid sched name?

File: src/scheduler/pbsfs.c (hunk 5 above)
Comment (bhroam) [reply src/scheduler/pbsfs.c]: I'll change it.  I'm not even sure it's a performance hit.  strncpy() will always copy N bytes.  If the argument is less than N bytes, it will fill out the rest of the N bytes with 0s.  Using snprintf() has the overhead of parsing the format string, but only copies what it needs to.

File: src/scheduler/pbsfs.c (hunk 7 above)
Comment (bhroam) [reply src/scheduler/pbsfs.c]: It turns out that add has been deprecated for some time.  I'm just going to remove it.  It's not in the 14.2.1 docs.

File: test/tests/functional/pbs_multi_sched.py (hunk 8 above)
Comment (bhroam) [reply test/tests/functional/pbs_multi_sched.py]: I'm a little confused on what you are asking me to test.  Are you asking me to try and create a scheduler with an invalid name?  I have no problem writing that test, but I didn't implement that code.  I'd have to familiarize myself with that portion of the code.  I did the multisched review, so it shouldn't be too hard.

File: test/fw/ptl/lib/pbs_testlib.py [hunk 9]
```diff
@@ -12084,11 +12109,11 @@ def decay_fairshare_tree(self):
         if self.has_diag:
             return True
 
-        self.stop()
         pbsfs = os.path.join(self.pbs_conf['PBS_EXEC'], 'sbin', 'pbsfs')
+        if self.sc_name is not 'default':
+            cmd += ['-I', self.sc_name]
```
Comment (visheshh) [inline test/fw/ptl/lib/pbs_testlib.py]: command is loaded to pbsfs here , So -I sched_name should be added to pbsfs.

File: test/fw/ptl/lib/pbs_testlib.py [hunk 10]
```diff
@@ -12118,6 +12143,8 @@ def cmp_fairshare_entities(self, name1=None, name2=None):
             name2 = str(name2)
 
         pbsfs = os.path.join(self.pbs_conf['PBS_EXEC'], 'sbin', 'pbsfs')
+        if self.sc_name is not 'default':
+            cmd += ['-I', self.sc_name]
```
Comment (visheshh) [inline test/fw/ptl/lib/pbs_testlib.py]: command is loaded to pbsfs here , So -I sched_name should be added to pbsfs , else move this if condition after next line

File: src/scheduler/pbsfs.c [hunk 11]
```diff
@@ -72,10 +72,9 @@ static void print_fairshare(group_info *root, int level);
 #define FS_PRINT 4
 #define FS_PRINT_TREE 8
 #define FS_DECAY 16
-#define FS_ADD 32
-#define FS_COMP 64
-#define FS_TRIM_TREE 128
-#define FS_WRITE_FILE 256
+#define FS_COMP 32
+#define FS_TRIM_TREE 64
+#define FS_WRITE_FILE 128
```
Comment (mike0042) [inline src/scheduler/pbsfs.c]: Technically, there's nothing wrong with defining these macros with decimal values. Probably better to express powers of two in hex or octal. Didn't see this in my previous review because it hadn't changed.

File: src/scheduler/pbsfs.c [hunk 12]
```diff
@@ -141,23 +141,58 @@ main(int argc, char *argv[])
 		exit(1);
 	}
 	if ((flags & (FS_PRINT | FS_PRINT_TREE)) && (argc - optind) != 0) {
-		fprintf(stderr, "Usage: pbsfs -[ptdgcs]\n");
+		fprintf(stderr, "Usage: pbsfs -[ptdgcs] [-I sched_name]\n");
 		exit(1);
 	}
 	else if ((flags & FS_GET)  && (argc - optind) != 1) {
-		fprintf(stderr, "Usage: pbsfs -g <fairshare_entity>\n");
+		fprintf(stderr, "Usage: pbsfs -i [sched_name] -g <fairshare_entity>\n");
```
Comment (latha-subramanian) [inline src/scheduler/pbsfs.c]: Shouldn't this be an upper case "I" ?

File: src/scheduler/pbsfs.c (hunk 11 above)
Comment (bhroam) [reply src/scheduler/pbsfs.c]: I'd rather leave them alone.  pbsfs is part of the scheduler, and the scheduler uses decimal powers of 2.  If I were to change these, I'd want to change the rest of the scheduler's defines.

File: test/fw/ptl/lib/pbs_testlib.py (hunk 9 above)
Comment (bhroam) [reply test/fw/ptl/lib/pbs_testlib.py]: oops, copy paste error.  Thanks for finding this.

File: src/scheduler/pbsfs.c [hunk 13]
```diff
@@ -141,23 +141,58 @@ main(int argc, char *argv[])
...
 		exit(1);
 	}
 	else if ((flags & FS_SET) && (argc - optind) != 2) {
-		fprintf(stderr, "Usage: pbsfs -s <fairshare_entity> <usage>\n");
+		fprintf(stderr, "Usage: pbsfs -I [sched_name] -s <fairshare_entity> <usage>\n");
```
Comment (arungrover) [inline src/scheduler/pbsfs.c]: @bhroam  shouldn't -I be inside the square brackets? right now it looks like -I is necessary but it may or may not need scheduler name

**Comment by bhroam** [comment]
I fixed the comments (except Mike's about decimal powers of 2).  I added 2 more tests to test the functions that @visheshh commented on.  They obviously weren't being tested.

**Comment by bhroam** [comment]
[ptl.log](https://github.com/PBSPro/pbspro/files/1738596/ptl.log)

File: src/scheduler/pbsfs.c (hunk 11 above)
Comment (mike0042) [reply src/scheduler/pbsfs.c]: Makes sense to me.

File: test/tests/functional/pbs_multi_sched.py (hunk 8 above)
Comment (mike0042) [reply test/tests/functional/pbs_multi_sched.py]: I was thinking about an illegal argument to the "-I" parameter. For example, if the scheduler "foo" did not exist then you should see the message "Scheduler foo does not exist" if you pass "-I foo". What do you think?

**Comment by mike0042** [comment]
Approved this PR. It's up to Bhroam whether he wants to add another test case. It's not mandatory.

**Comment by latha-subramanian** [comment]
@Bhroam,  The changes LGTM. I sign off

File: src/scheduler/pbsfs.c [hunk 14]
```diff
@@ -141,23 +141,58 @@ main(int argc, char *argv[])
...
 		exit(1);
 	}
 
-	sprintf(path_buf, "%s/sched_priv/", pbs_conf.pbs_home_path);
+	if (strcmp(sched_name, "default") != 0) {
+		int pbs_sd;
+		struct batch_status *bs;
+		struct batch_status *cur_bs;
+		pbs_sd = pbs_connect(NULL);
+		if(pbs_sd < 0) {
```
Comment (vstumpf) [inline src/scheduler/pbsfs.c]: Should be a space after if if

**Comment by bhroam** [comment]
Squashed and rebased.  @PBSPro/pbspro-maintainers please take this forward.

File: test/tests/functional/pbs_multi_sched.py (hunk 8 above)
Comment (bhroam) [reply test/tests/functional/pbs_multi_sched.py]: I think I'll skip this test.  It would require me to go outside of PTL's fairshare support and run pbsfs directly.  While I can do it, I'd rather not.

**Comment by mike0042** [comment]
Restarted the TravisCI builds.

**Comment by mike0042** [comment]
There seems to be a problem with TravisCI at the moment. All of the builds are failing and the reason is unclear other than the error is coming from ./github/runchecks

**Comment by mike0042** [comment]
I see why... please sign your commit.

**Comment by bhroam** [comment]
Sorry about that.  I have autosign set up on my VM, but git on my VM wasn't playing nice with my branch.  I needed to squash with git on my mac.  I forgot to go back and sign it.  It's signed now.

**Comment by subhasisb** [comment]
@bhroam I think we need both for now. It would probably help to go ahead with "", but also file a bug for windows so that somebody can investigate/fix this issue in general on windows. Thanks.

**Comment by bhroam** [comment]
Marked PR as [WIP] due to windows issues.  I'll try changing from pbs_connect(NULL) to pbs_connect("").  This should be consistent with what the commands do.

**Comment by mike0042** [comment]
Latest change looks fine.

**Comment by mike0042** [comment]
Still looks good.

**Comment by bhroam** [comment]
Adding the call to winsock_init() fixed the windows problem.  I've removed [WIP].  @vstumpf if you could look at the last few commits and sign off again, I'll squash and rebase.

**Comment by bhroam** [comment]
Thanks for the sign offs.  I just squashed and rebased.
